### PR TITLE
Implemented bunch of Utf8String/Span APIs

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
@ahsonkhan, @GrabYourPitchforks, @joshfree, @A-And 

These are not very optimized, but I am just making sure we have the basics working.

### CODE REVIEW

File: src/System.Text.Utf8String/System/Text/Utf8Span.cs [hunk 1]
```diff
@@ -508,6 +513,17 @@ private static int LastIndexOf(ReadOnlySpan<byte> buffer, byte value)
             }
             return -1;
         }
+
+        private int Compare(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right)
```
Comment (joshfree) [inline src/System.Text.Utf8String/System/Text/Utf8Span.cs]: I know you said this isn't optimized yet, but can you look at reusing some of https://github.com/dotnet/coreclr/blob/8ec9b34df095df032e37b4c20a7359e9df7ae0ce/src/mscorlib/src/System/String.Comparison.cs#L222

Comment (KrzysztofCwalina) [reply src/System.Text.Utf8String/System/Text/Utf8Span.cs]: I filed issue https://github.com/dotnet/corefx/issues/24880 and added comment to borrow from this implementation.

Comment (joshfree) [reply src/System.Text.Utf8String/System/Text/Utf8Span.cs]: Thanks

File: src/System.Text.Utf8String/System/Text/Utf8String.cs [hunk 2]
```diff
@@ -285,14 +295,15 @@ public Utf8String TrimEnd(Utf8String trimCharacters)
 
         // TODO: should we even have index based operations?
         #region Index-based operations
-        public Utf8String Substring(int index) => Substring(index, Bytes.Length - index);
+        public Utf8String Substring(int index) => index==0?this:Substring(index, Bytes.Length - index);
```
Comment (ahsonkhan) [inline src/System.Text.Utf8String/System/Text/Utf8String.cs]: nit: fix spacing here and elsewhere

File: src/System.Text.Utf8String/System/Text/Utf8Span.cs [hunk 3]
```diff
@@ -508,6 +513,17 @@ private static int LastIndexOf(ReadOnlySpan<byte> buffer, byte value)
...
+        {
+            var minLength = left.Length;
+            if(minLength > right.Length) minLength = right.Length;
+            for(int i=0; i<minLength; i++) {
```
Comment (ahsonkhan) [inline src/System.Text.Utf8String/System/Text/Utf8Span.cs]: nit: spacing to improve readability 
```C#
for (int i = 0; i < minLength; i++) { 
```

File: src/System.Text.Utf8String/System/Text/Utf8String.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply src/System.Text.Utf8String/System/Text/Utf8String.cs]: thanks. will fix in the next PR

File: src/System.Text.Utf8String/System/Text/Utf8Span.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply src/System.Text.Utf8String/System/Text/Utf8Span.cs]: thanks. will fix

**Pull request merged**